File: cpp/src/smb.cpp
#include "nesle/smb.hpp"

#include <array>
#include <stdexcept>

namespace nesle::smb {
namespace {

int read_digits(std::span<const std::uint8_t> ram, std::size_t address, std::size_t length) {
    int value = 0;
    for (std::size_t i = 0; i < length; ++i) {
        value = value * 10 + static_cast<int>(ram[address + i] & 0x0F);
    }
    return value;
}

bool is_stage_over(std::span<const std::uint8_t> ram) {
    constexpr std::array<std::uint8_t, 2> kStageOverEnemies = {0x2D, 0x31};
    for (std::size_t i = 0; i < 5; ++i) {
        const auto enemy = ram[kEnemyTypeBase + i];
        if ((enemy == kStageOverEnemies[0] || enemy == kStageOverEnemies[1]) &&
            ram[kPlayerFloatState] == 3) {
            return true;
        }
    }
    return false;
}

}  // namespace

MarioRamState read_ram(std::span<const std::uint8_t> ram) {
    if (ram.size() < kCpuRamBytes) {
        throw std::invalid_argument("Super Mario Bros. RAM view must contain 2048 bytes");
    }

    MarioRamState state;
    state.x_pos = static_cast<int>(ram[kXPage]) * 0x100 + static_cast<int>(ram[kXScreen]);
    state.y_viewport = ram[kYViewport];
    state.y_pos = state.y_viewport < 1 ? 255 + (255 - static_cast<int>(ram[kYPixel]))
                                       : 255 - static_cast<int>(ram[kYPixel]);
    state.time = read_digits(ram, kTimeDigits, 3);
    state.coins = read_digits(ram, kCoinsDigits, 2);
    state.score = read_digits(ram, kScoreDigits, 6);
    state.lives = ram[kLives];
    state.world = static_cast<int>(ram[kWorld]) + 1;
    state.stage = static_cast<int>(ram[kStage]) + 1;
    state.area = static_cast<int>(ram[kArea]) + 1;
    state.status_code = ram[kStatus];
    state.player_state = ram[kPlayerState];
    state.is_dying = state.player_state == 0x0B || state.y_viewport > 1;
    state.is_dead = state.player_state == 0x06;
    state.is_game_over = state.lives == 0xFF;
    state.flag_get = ram[kGameMode] == 2 || is_stage_over(ram);
    return state;
}

RewardComponents compute_reward(const MarioRamState& previous, const MarioRamState& current) {
    RewardComponents reward;
    reward.x = current.x_pos - previous.x_pos;
    if (reward.x < -5 || reward.x > 5) {
        reward.x = 0;
    }

    reward.time = current.time - previous.time;
    if (reward.time > 0) {
        reward.time = 0;
    }

    reward.death = (current.is_dying || current.is_dead) ? -25 : 0;
    reward.total = reward.x + reward.time + reward.death;
    return reward;
}

std::string status_name(int status_code) {
    if (status_code == 0) {
        return "small";
    }
    if (status_code == 1) {
        return "tall";
    }
    return "fireball";
}

std::string implausible_boot_state_reason(const MarioRamState& state) {
    if (state.world < 1 || state.world > 8) {
        return "Mario world RAM is outside the expected 1..8 range";
    }
    if (state.stage < 1 || state.stage > 4) {
        return "Mario stage RAM is outside the expected 1..4 range";
    }
    if (state.area < 1 || state.area > 32) {
        return "Mario area RAM is outside the expected 1..32 range";
    }
    if (state.time < 0 || state.time > 999) {
        return "Mario timer RAM is outside the expected 0..999 range";
    }
    if (state.coins < 0 || state.coins > 99) {
        return "Mario coin RAM is outside the expected 0..99 range";
    }
    if (state.lives > 99 && state.lives != 0xFF) {
        return "Mario lives RAM is outside the expected range";
    }
    if (state.status_code < 0 || state.status_code > 2) {
        return "Mario power-up status RAM is outside the expected 0..2 range";
    }
    if (state.player_state > 0x0D) {
        return "Mario player-state RAM is outside the expected range";
    }
    if (state.x_pos == 0 && state.time == 0 && state.lives == 0 &&
        state.world == 1 && state.stage == 1 && state.area == 1) {
        return "Mario RAM still looks like an uninitialized all-zero boot state";
    }
    return "";
}

bool is_plausible_boot_state(const MarioRamState& state) {
    return implausible_boot_state_reason(state).empty();
}

void validate_plausible_boot_state(const MarioRamState& state) {
    const auto reason = implausible_boot_state_reason(state);
    if (!reason.empty()) {
        throw std::invalid_argument(reason);
    }
}

}  // namespace nesle::smb

File: src/nesle/smb.py
from __future__ import annotations

from dataclasses import dataclass


CPU_RAM_BYTES = 2048

PLAYER_STATE = 0x000E
PLAYER_FLOAT_STATE = 0x001D
ENEMY_TYPE_BASE = 0x0016
X_PAGE = 0x006D
X_SCREEN = 0x0086
Y_VIEWPORT = 0x00B5
Y_PIXEL = 0x03B8
LIVES = 0x075A
STAGE = 0x075C
WORLD = 0x075F
AREA = 0x0760
STATUS = 0x0756
GAME_MODE = 0x0770
COINS_DIGITS = 0x07ED
TIME_DIGITS = 0x07F8
SCORE_DIGITS = 0x07DE


@dataclass(frozen=True)
class MarioRamState:
    x_pos: int
    y_pos: int
    time: int
    coins: int
    score: int
    life: int
    world: int
    stage: int
    area: int
    status: str
    status_code: int
    player_state: int
    flag_get: bool
    is_dying: bool
    is_dead: bool
    is_game_over: bool


@dataclass(frozen=True)
class RewardComponents:
    x: int
    time: int
    death: int
    total: int


def _read_digits(ram: bytes, address: int, length: int) -> int:
    value = 0
    for byte in ram[address : address + length]:
        value = value * 10 + (byte & 0x0F)
    return value


def _status_name(status_code: int) -> str:
    if status_code == 0:
        return "small"
    if status_code == 1:
        return "tall"
    return "fireball"


def _is_stage_over(ram: bytes) -> bool:
    stage_over_enemies = {0x2D, 0x31}
    for offset in range(5):
        if ram[ENEMY_TYPE_BASE + offset] in stage_over_enemies and ram[PLAYER_FLOAT_STATE] == 3:
            return True
    return False


def read_ram(data: bytes | bytearray | memoryview) -> MarioRamState:
    ram = bytes(data)
    if len(ram) < CPU_RAM_BYTES:
        raise ValueError("Super Mario Bros. RAM view must contain 2048 bytes")

    y_viewport = ram[Y_VIEWPORT]
    y_pixel = ram[Y_PIXEL]
    player_state = ram[PLAYER_STATE]
    status_code = ram[STATUS]

    return MarioRamState(
        x_pos=ram[X_PAGE] * 0x100 + ram[X_SCREEN],
        y_pos=255 + (255 - y_pixel) if y_viewport < 1 else 255 - y_pixel,
        time=_read_digits(ram, TIME_DIGITS, 3),
        coins=_read_digits(ram, COINS_DIGITS, 2),
        score=_read_digits(ram, SCORE_DIGITS, 6),
        life=ram[LIVES],
        world=ram[WORLD] + 1,
        stage=ram[STAGE] + 1,
        area=ram[AREA] + 1,
        status=_status_name(status_code),
        status_code=status_code,
        player_state=player_state,
        flag_get=ram[GAME_MODE] == 2 or _is_stage_over(ram),
        is_dying=player_state == 0x0B or y_viewport > 1,
        is_dead=player_state == 0x06,
        is_game_over=ram[LIVES] == 0xFF,
    )


def compute_reward(previous: MarioRamState, current: MarioRamState) -> RewardComponents:
    x_reward = current.x_pos - previous.x_pos
    if x_reward < -5 or x_reward > 5:
        x_reward = 0

    time_reward = current.time - previous.time
    if time_reward > 0:
        time_reward = 0

    death_penalty = -25 if current.is_dying or current.is_dead else 0
    return RewardComponents(
        x=x_reward,
        time=time_reward,
        death=death_penalty,
        total=x_reward + time_reward + death_penalty,
    )
